datafusion/core/src/physical_optimizer/sort_pushdown.rs

```diff
@@ -0,0 +1,416 @@
+            }
+        })
+        .collect()
+    }
+}
+
+impl TreeNode for SortPushDown {
+    fn apply_children<F>(&self, op: &mut F) -> Result<VisitRecursion>
+    where
+        F: FnMut(&Self) -> Result<VisitRecursion>,
+    {
+        let children = self.children();
+        for child in children {
+            match op(&child)? {
+                VisitRecursion::Continue => {}
+                VisitRecursion::Skip => return Ok(VisitRecursion::Continue),
+                VisitRecursion::Stop => return Ok(VisitRecursion::Stop),
+            }
+        }
+
+        Ok(VisitRecursion::Continue)
+    }
+
+    fn map_children<F>(mut self, transform: F) -> Result<Self>
+    where
+        F: FnMut(Self) -> Result<Self>,
+    {
+        let children = self.children();
+        if !children.is_empty() {
+            let children_plans = children
+                .into_iter()
+                .map(transform)
+                .map(|r| r.map(|s| s.plan))
+                .collect::<Result<Vec<_>>>()?;
+
+            match with_new_children_if_necessary(self.plan, children_plans)? {
+                Transformed::Yes(plan) | Transformed::No(plan) => {
+                    self.plan = plan;
+                }
+            }
+        };
+        Ok(self)
+    }
+}
+
+pub(crate) fn pushdown_sorts(
+    requirements: SortPushDown,
+) -> Result<Transformed<SortPushDown>> {
+    let plan = &requirements.plan;
+    let parent_required = requirements.required_ordering.as_deref();
+    const ERR_MSG: &str = "Expects parent requirement to contain something";
+    let err = || DataFusionError::Plan(ERR_MSG.to_string());
+    if let Some(sort_exec) = plan.as_any().downcast_ref::<SortExec>() {
+        let mut new_plan = plan.clone();
+        if !ordering_satisfy_requirement(plan.output_ordering(), parent_required, || {
+            plan.equivalence_properties()
+        }) {
+            // If the current plan is a SortExec, modify it to satisfy parent requirements:
+            let parent_required_expr =
+                make_sort_exprs_from_requirements(parent_required.ok_or_else(err)?);
+            new_plan = sort_exec.input.clone();
+            add_sort_above(&mut new_plan, parent_required_expr)?;
+        };
+        let required_ordering = new_plan
+            .output_ordering()
+            .map(make_sort_requirements_from_exprs);
+        // Since new_plan is a SortExec, we can safely get the 0th index.
+        let child = &new_plan.children()[0];
+        if let Some(adjusted) =
+            pushdown_requirement_to_children(child, required_ordering.as_deref())?
+        {
+            // Can push down requirements
+            Ok(Transformed::Yes(SortPushDown {
+                plan: child.clone(),
+                required_ordering,
+                adjusted_request_ordering: adjusted,
+            }))
+        } else {
+            // Can not push down requirements
+            Ok(Transformed::Yes(SortPushDown::init(new_plan)))
+        }
```

Comment:
I have one question for this logic, since the parent requirements are already satisfied by modifying the SortExec, why you need to push down the requirements here ?

Reply:
Actually, you are right. No need for it. Changed it to `None`.